src/PairedRouting.php

```diff
@@ -772,6 +772,38 @@ public function add_permalink_settings_notice() {
 		<?php
 	}
 
+	/**
+	 * Update canonical URL for legacy reader paired urls.
+	 *
+	 * @param string $redirect_url Canonical redirect URL.
+	 *
+	 * @return string Updated canonical URL.
+	 */
+	public function maybe_update_paired_url( $redirect_url ) {
+
+		global $wp_rewrite;
+
+		/**
+		 * To address issue AMP version of a comment page.
+		 * Reference: https://github.com/ampproject/amp-wp/issues/6355
+		 */
+		if ( ! empty( $redirect_url ) &&
+			! empty( $wp_rewrite->comments_pagination_base ) &&
+			AMP_Options_Manager::get_option( Option::PAIRED_URL_STRUCTURE ) === Option::PAIRED_URL_STRUCTURE_LEGACY_READER
```

Comment:
I condition is not required:

```suggestion
			! empty( $wp_rewrite->comments_pagination_base )
```

The reason is that `maybe_update_paired_url` is only added if `is_using_path_suffix` and I think the problem is present for both Legacy Reader and Path Suffix paired URL structures, although only the former results in a 404.

Reply:
Interestingly, when the Path Suffix paired URL structure is active, no redirection happens although the `redirect_canonical` filter is called first with:

```json
{
    "redirect_url": "https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/amp/comment-page-2/amp/",
    "requested_url": "https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/amp/"
}
```

And then again with:

```json
{
    "redirect_url": "https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/amp/comment-page-2/amp/comment-page-2/amp/",
    "requested_url": "https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/amp/comment-page-2/amp/"
}
```

Nevertheless, no redirection happens.

In contrast, when the Legacy Reader paired URL structure is chosen, then I see the `redirect_canonical` filter called with:

```json
{
    "redirect_url": "https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/amp/comment-page-2/?amp",
    "requested_url": "https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/amp/"
}
```

And I am then redirected to https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/amp/comment-page-2/?amp

Which the AMP plugin then redirects me to https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/amp/comment-page-2/

And this then results in a 404.

Reply:
That is, without this patch applied.

Nevertheless, with this patch I am seeing a redirection happen when the Legacy Reader paired URL structure is active:

```
https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/amp/
=>
https://wordpressdev.lndo.site/2012/01/03/template-comments/comment-page-2/?amp
```

But this is not expected.

Reply:
Fixed in acf8a172c7173a361d92d0407170bfd8f9a62d8a.